disnake/i18n.py

```diff
@@ -0,0 +1,280 @@
+DEALINGS IN THE SOFTWARE.
+"""
+
+from __future__ import annotations
+
+import logging
+import os
+import warnings
+from abc import ABC, abstractmethod
+from collections import defaultdict
+from pathlib import Path
+from typing import DefaultDict, Dict, Optional, Set, Union
+
+from . import utils
+from .custom_warnings import LocalizationWarning
+from .enums import Locale
+
+MISSING = utils.MISSING
+
+
+Localizations = Union[str, Dict[Locale, str], Dict[str, str]]
+
+
+__all__ = (
+    "LocalizationProtocol",
+    "LocalizationStore",
+)
+
+
+_log = logging.getLogger(__name__)
+
+
+class LocalizationValue:
+    """
+    Internal container type for (pending) localization data
+
+    .. versionadded:: 2.5
+    """
+
+    __slots__ = ("_key", "_data")
+
+    def __init__(self, localizations: Optional[Localizations]):
+        self._key: Optional[str]
+        self._data: Optional[Dict[str, str]]
+
+        if localizations is None:
+            # no localization
+            self._key = None
+            self._data = None
+        elif isinstance(localizations, str):
+            # got localization key
+            self._key = localizations
+            self._data = MISSING  # not localized yet
+        elif isinstance(localizations, dict):
+            # got localization data
+            self._key = None
+            self._data = {str(k): v for k, v in localizations.items()}
+        else:
+            raise TypeError(f"Invalid localizations type: {type(localizations).__name__}")
+
+    def _link(self, store: LocalizationProtocol) -> None:
+        """Loads localizations from the specified store if this object has a key"""
+        if self._key is not None:
+            self._data = store.get(self._key)
+
+    def to_dict(self) -> Optional[Dict[str, str]]:
+        """Returns a dict with a locale -> localization mapping"""
+        if self._data is MISSING:
+            warnings.warn(
+                "value was never localized, this is likely a library bug",
+                LocalizationWarning,
+                stacklevel=2,
+            )
+            return None
+        return self._data
+
+    def __eq__(self, other) -> bool:
+        d1 = self.to_dict()
+        d2 = other.to_dict()
+        # consider values equal if they're both falsy, or actually equal
+        # (it doesn't matter if localizations are `None` or `{}`)
+        return (not d1 and not d2) or d1 == d2
+
+
+class LocalizationProtocol(ABC):
+    """
+    Manages a key-value mapping of localizations.
+
+    This is an abstract class, a concrete implementation is provided as :class:`LocalizationStore`.
+
+    .. versionadded:: 2.5
+    """
+
+    @abstractmethod
+    def get(self, key: str) -> Optional[Dict[str, str]]:
+        """
+        Returns localizations for the specified key.
+
+        Parameters
+        ----------
+        key: :class:`str`
+            The lookup key.
+
+        Raises
+        ------
+        RuntimeError
```

Comment:
We should have a specific exception, instead of a RuntimeError

Reply:
78a7378dcd2d198cb02346bb44dbaa9747f08c66